kernel/posix/pthread.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <atomic.h>
+#include "ksched.h"
+#include "wait_q.h"
+
+/* Kernel Structure for Number of threads */
+struct k_thread threads[CONFIG_MAX_PTHREAD_COUNT];
+
+static int is_prio_valid(int priority)
+{
+	if ((priority < K_HIGHEST_APPLICATION_THREAD_PRIO)
+	    || (priority > K_HIGHEST_APPLICATION_THREAD_PRIO)) {
+		return -1;
+	} else       {
+		return 0;
+	}
+
+}
+
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	/*
+	 *  Task priorities can range from  K_LOWEST_THREAD_PRIO to
+	 *  K_HIGHEST_THREAD_PRIO.
+	 */
+	if (is_prio_valid(priority) != 0) {
+		/* Invalid Prioritty */
+		errno = EINVAL;
+		return -1;
+	}
+	attr->priority = priority;
+	return 0;
+}
+
+int pthread_attr_setstack(pthread_attr_t *attr, void *stackaddr,
+			  size_t stacksize)
+{
+
+	if ((stackaddr == NULL) || (stacksize <= K_LOWEST_STACK_SIZE)) {
+		errno = EINVAL;
+		return -1;
+	}
+	attr->stack = stackaddr;
+	attr->stacksize = stacksize;
+	return 0;
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	void * (*fun_ptr)(void *);
+
+	fun_ptr = arg3;
+	(*fun_ptr)((void *)arg1);
+}
+
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static int thread_count;
+
+	if ((attr == NULL) || (attr->stack == NULL)) {
+		errno = EINVAL;
+		return -1;
+	}
+	if (thread_count < CONFIG_MAX_PTHREAD_COUNT) {
```

Comment:
Between creating a thread and incrementing the value, preemption might occur, causing a race condition that will lead to invalid memory access.

`atomic_inc()` will return the previous value, which is the one you want to use:

```C

atomic_val_t index = atomic_inc(&thread_count);
if (index < CONFIG_MAX_PTHREAD_COUNT) {
    ...  &threads[index] ...
}
```

This way, `index` is guaranteed to be below `CONFIG_MAX_PTHREAD_COUNT`.  This way, there's no race condition: there's no way to have an element in the threads array that's going to be skipped, or two threads that are initialized in the same element.

Reply:
Thanks @lpereira I will take care of this.